iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp

```diff
@@ -0,0 +1,353 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+#define DEBUG_TYPE "iree-flow-canonicalize-dispatch-ids"
+
+/// This file contains patterns allowing canonicalization taking advantage  of
+/// flow dispatch ops. When the ID range is know we may be able to calculate the
+/// min/max of the derived values and may allow folding affine.min or ForOp.
+/// This file duplicate some of the logic from
+/// `linalg::AffineMinSCFCanonicalizationPattern`. In the future we could unify
+/// this code back by upstreaminng an interface to MLIR to represent generic ID
+/// ops.
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+/// Traverse `e` and return an AffineExpr where all occurrences of `dim` have
+/// been replaced by either:
+///  - `min` if `positivePath` is true when we reach an occurrence of `dim`
+///  - `max` if `positivePath` is true when we reach an occurrence of `dim`
+/// `positivePath` is negated each time we hit a multiplicative or divisive
+/// binary op with a constant negative coefficient.
+static AffineExpr substWithMin(AffineExpr e, AffineExpr dim, AffineExpr min,
+                               AffineExpr max, bool positivePath = true) {
+  if (e == dim) return positivePath ? min : max;
+  if (auto bin = e.dyn_cast<AffineBinaryOpExpr>()) {
+    AffineExpr lhs = bin.getLHS();
+    AffineExpr rhs = bin.getRHS();
+    if (bin.getKind() == mlir::AffineExprKind::Add)
+      return substWithMin(lhs, dim, min, max, positivePath) +
+             substWithMin(rhs, dim, min, max, positivePath);
+
+    auto c1 = bin.getLHS().dyn_cast<AffineConstantExpr>();
```

Comment:
The below holds only for `AffineExprKind::Mul`? What about `FloorDiv` and `CeilDiv`? I would assume that when `positivePath=true` the denominator must be substituted with the max value.

Reply:
Below holds for Mul, Mod, FloorDiv and CeilDiv (everything but Add). Yes if the denominator is a negative constant we need to swap the `positivePath` value. This is what is being done.

Reply:
As I understand what is happening here, you want to take an expression and find its "minimum" based on the dim being between a certain range right?

so if `d0` is in range `[1, 50]` and you have `e / d0` wont min of this expression be if you replace `d0` with its `max` value ? i.e. `e / 50`? Maybe I am missing the intent

Reply:
Hmmm, as I write that, i realize that `e/d0` cant happen. Its not affine. So its OK. Ill take a look again